akka-remote/src/main/scala/akka/remote/artery/Codecs.scala

```diff
@@ -224,6 +224,99 @@ private[remote] object Decoder {
     inboundEnvelope: InboundEnvelope)
 
   private object Tick
+
+  /** Materialized value of [[Encoder]] which allows safely calling into the stage to interfact with compression tables. */
+  private[remote] trait InboundCompressionAccess {
+    def confirmActorRefCompressionAdvertisementAck(ack: ActorRefCompressionAdvertisementAck): Future[Done]
+    def confirmClassManifestCompressionAdvertisementAck(ack: ClassManifestCompressionAdvertisementAck): Future[Done]
+    def closeCompressionFor(originUid: Long): Future[Done]
+
+    /** For testing purposes, usually triggered by timer from within Decoder stage. */
+    def runNextActorRefAdvertisement(): Unit
+    /** For testing purposes, usually triggered by timer from within Decoder stage. */
+    def runNextClassManifestAdvertisement(): Unit
+  }
+
+  private[remote] trait InboundCompressionAccessImpl extends InboundCompressionAccess {
+    this: GraphStageLogic with StageLogging ⇒
+
+    def compressions: InboundCompressions
+
+    private val closeCompressionForCb = getAsyncCallback[(Long, Promise[Done])] {
+      case (uid, done) ⇒
+        compressions.close(uid)
+        done.success(Done)
+    }
+    private val confirmActorRefCompressionAdvertisementCb = getAsyncCallback[(ActorRefCompressionAdvertisementAck, Promise[Done])] {
+      case (ActorRefCompressionAdvertisementAck(from, tableVersion), done) ⇒
+        compressions.confirmActorRefCompressionAdvertisement(from.uid, tableVersion)
+        done.success(Done)
+    }
+    private val confirmClassManifestCompressionAdvertisementCb = getAsyncCallback[(ClassManifestCompressionAdvertisementAck, Promise[Done])] {
+      case (ClassManifestCompressionAdvertisementAck(from, tableVersion), done) ⇒
+        compressions.confirmClassManifestCompressionAdvertisement(from.uid, tableVersion)
+        done.success(Done)
+    }
+    private val runNextActorRefAdvertisementCb = getAsyncCallback[Unit] {
+      _ ⇒ compressions.runNextActorRefAdvertisement()
+    }
+    private val runNextClassManifestAdvertisementCb = getAsyncCallback[Unit] {
+      _ ⇒ compressions.runNextClassManifestAdvertisement()
+    }
+
+    // TODO in practice though all those CB's will always succeed, no need for the futures etc IMO
```

Comment:
second opinion on that one? The callbacks in the Encoder are wrapped like that since it is initialized later, however in the Decoder once constructed we're ready to go so I don't see how the invokations could fail.
On the other hand, it's not hot path so it may be safer this way in case something goes wrong.
We don't use the futures anywhere though. So I was tempted to not use any of the Future dance (see like 274 for an example)